component/src/mock_client.rs

```diff
@@ -75,8 +78,39 @@ impl MockClient {
                         }
                     }
                 }
+                StatePayload::Swap { swap: payload, .. } => {
+                    match payload.trial_decrypt(&self.fvk) {
+                        Some(swap) => {
+                            self.nct.insert(Keep, payload.commitment)?;
+                            // At this point, we need to retain the swap plaintext,
+                            // and also derive the expected output notes so we can
+                            // notice them while scanning later blocks.
+                            self.swaps.insert(payload.commitment, swap.clone());
+
+                            let batch_data =
+                                block.swap_outputs.get(&swap.trading_pair).ok_or_else(|| {
+                                    anyhow::anyhow!("server gave invalid compact block")
+                                })?;
+
+                            let (output_1, output_2) = swap.output_notes(batch_data);
+                            // Pre-insert the output notes into our notes table, so that
+                            // we can notice them when we scan the block where they are claimed.
+                            self.notes.insert(output_1.commit(), output_1);
+                            self.notes.insert(output_2.commit(), output_2);
+                        }
+                        None => {
+                            self.nct.insert(Forget, payload.commitment)?;
+                        }
+                    }
+                }
```

Comment:
The `MockClient` does double duty, also serving as a worked example of the scanning logic without any fancy optimizations, and here is the logic for scanning `Swap`s: after trial decryption, grab the output data that's bundled into the compact block, and use it to pre-insert the notes that will be generated in the claim, ...